### PR TITLE
Changes to make CUDA features work in Dask, Distributed and Spark

### PR DESCRIPTION
**Pull request by sklam**
- [x] make ``cuda.jit`` and cuda ufunc picklable
- [x] detect forked after cuda driver init
- [x] resolve CUDA driver issues with multiprocess/multithreaded
- [x] examples

Compatibility status:
- [x] dask
- [x] distributed
- [x] spark

### CODE REVIEW

**Comment by seibert** [comment]
Are `@vectorize` ufuncs for the other targets (cpu and parallel) picklable?

**Comment by sklam** [comment]
@seibert , cpu/parallel targets are harder due to dependency with numpy ufunc machinery.  The CUDA ufunc is a custom object that is ufunc-like.  The changes here are mostly ensuring CUDA jit-ed functions are picklable; then, the CUDA ufunc object is automatically picklable.  The changes (https://github.com/sklam/numba/commit/c3bf39bed8127237c421405830dc66a9e75cb909#diff-aa526021a7f3212f14d02c75535106afR27)  is for avoiding CUDA driver initialization at when creating a CUDA ufunc.

**Comment by pitrou** [comment]
I don't like the sprinkling of LLVM locks here and there. Can't this be centralized in the codegen module? The fact that we're parsing LLVM IR by hand elsewhere is not very pretty...

File: numba/cuda/compiler.py [hunk 1]
```diff
@@ -1,18 +1,21 @@
...
 from numba.typing.templates import ConcreteTemplate
-from numba import funcdesc, typing, utils
+from numba import funcdesc, typing, utils, serialize
 
 from .cudadrv.autotune import AutoTuner
 from .cudadrv.devices import get_context
 from .cudadrv import nvvm, devicearray, driver
 from .errors import KernelRuntimeError
 from .api import get_current_device
+from functools import reduce
```
Comment (pitrou) [inline numba/cuda/compiler.py]: Hmm, just a detail but can you leave the order of imports consistent? :-) (first stdlib then third-party then package-wide then local imports).

File: numba/cuda/tests/cudapy/test_serialize.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
+from __future__ import print_function
+import pickle
+import numpy as np
+from numba import cuda, vectorize
+from numba import unittest_support as unittest
+from numba.config import ENABLE_CUDASIM
+
+
+@unittest.skipIf(ENABLE_CUDASIM, 'pickling not supported in CUDASIM')
```
Comment (pitrou) [inline numba/cuda/tests/cudapy/test_serialize.py]: Why not `skip_on_cudasim`?

File: numba/cuda/tests/cudapy/test_serialize.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    def test_pickling_autojit(self):
+
+        @cuda.jit(device=True)
+        def inner(a):
+            return a + 1
+
+        @cuda.jit
+        def foo(arr):
+            arr[0] = inner(arr[0])
+
```
Comment (pitrou) [inline numba/cuda/tests/cudapy/test_serialize.py]: Did you forget a call to `self.check_call` in this test?

**Comment by pitrou** [comment]
As a sidenote, this patch seems to only protect *some* calls to LLVM, which sounds wrong (for example calls to `verify()` don't seem protected).

**Comment by sklam** [comment]
Perhaps, I should look at enabling threadsafe llvm.  According to http://llvm.org/docs/CMake.html, ``LLVM_ENABLE_THREADS`` is "ON" by default.  We just need to add a few API to llvmlite.

**Comment by sklam** [comment]
Just realize we are already in threadsafe mode.  The problem is the LLVMContext is never threadsafe.  Seems like rust compiler is threadsafe by letting each thread has its own LLVMContext.  I will give it a try.

**Comment by pitrou** [comment]
What does a LLVM context do/store?

**Comment by sklam** [comment]
The description (http://llvm.org/docs/doxygen/html/classllvm_1_1LLVMContext.html) saids:


> This is an important class for using LLVM in a threaded context.
>
> It (opaquely) owns and manages the core "global" data of LLVM's core infrastructure, including the type and constant uniquing tables. LLVMContext itself provides no locking guarantees, so you should be careful to have one context per thread.

**Comment by sklam** [comment]
I have hacked up some code that lazy-initialize a LLVMContext for each thread in a TLS.  But, per-thread LLVMContext has a few challenges:
1. can't link two modules from difficult context
2. workaround for 1 is to copy the module for the current thread/context, but this means higher memory overhead.
3. no easy way to properly clean up memory in 2; esp, when the application makes and jit in many threads.

File: numba/cuda/tests/cudapy/test_serialize.py (hunk 3 above)
Comment (sklam) [inline numba/cuda/tests/cudapy/test_serialize.py]: Fixed at 5a63888

**Comment by sklam** [comment]
Using finegrain locking at codegen.py is not possible.  The ``llvmlite.binding.Module.__del__`` can occur at arbitrary locations.  It's best to apply the locking at top-level compiler entry points that contain all usage of ``llvmlite.binding.*``.

**Comment by sklam** [comment]
What I have so far does not fix parallel compilation that involve multiple targets (e.g. CPU+CUDA).  Here, I want to propose two approaches:

1. Use a `numba` wide compiler lock to disable parallel usage of anything in LLVM;
2. Apply per-target locking and per-target LLVMContext, so that a CPU compilation and a CUDA compilation can happen at the same time.  This is help isolate any LLVMContext-wide data among targets.

Thoughts?

**Pull request merged**